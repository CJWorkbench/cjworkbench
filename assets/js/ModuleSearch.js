/**
* Search field that returns modules matching text input.
* 
*/

import React from 'react';
import Autosuggest from 'react-autosuggest';
import PropTypes from 'prop-types'
import { DragSource } from 'react-dnd';

const spec = {
  beginDrag(props, monitor, component) {
    return {
      index: false,
      id: props.id,
      insert: true,
    }
  },
  endDrag(props, monitor, component) {
    if (monitor.didDrop()) {
      const result = monitor.getDropResult();
      props.dropModule(result.source.id, result.source.index);
    }
  }
}

function collect(connect, monitor) {
  return {
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging()
  }
}

class ModuleSearchResult extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {
    return this.props.connectDragSource(
      <div className='react-autosuggest__suggestion-inner'>
      <div className='suggest-handle'>
        <div className='icon-grip'></div>
      </div>
        <div className='d-flex align-items-center'>
          <span className={'ml-icon-search icon-' + this.props.icon}></span>
          <span className='mt-1 content-3'><strong>{this.props.name}</strong></span>
        </div>
      </div>
    )
  }
}

const DraggableModuleSearchResult = DragSource('module', spec, collect)(ModuleSearchResult)

export default class ModuleSearch extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: '',
      suggestions: [],
      modules: []
    };
    this.onChange = this.onChange.bind(this);
    this.escapeRegexCharacters = this.escapeRegexCharacters.bind(this);
    this.onSuggestionsClearRequested = this.onSuggestionsClearRequested.bind(this);
    this.onSuggestionsFetchRequested = this.onSuggestionsFetchRequested.bind(this);
    this.renderSectionTitle = this.renderSectionTitle.bind(this);
    this.formatModules = this.formatModules.bind(this);
    this.getSuggestions = this.getSuggestions.bind(this);
    this.renderSuggestion = this.renderSuggestion.bind(this);
    this.getSuggestionValue = this.getSuggestionValue.bind(this);
    this.getSectionSuggestions = this.getSectionSuggestions.bind(this);
    this.onSuggestionSelected = this.onSuggestionSelected.bind(this);
  }

  componentDidMount() {
    if (this.props.items) {
      this.formatModules(this.props.items);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.items !== nextProps.items) {
      this.formatModules(nextProps.items);
    }
  }

  escapeRegexCharacters (str) {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  }
  formatModules(items) {
    let modules = [];
    let temp = {};

    if (items.length) {
      items.forEach(item => {
        if (temp[item.category]) {
          temp[item.category].push(item);
        } else {
          temp[item.category] = [item];
        }
      });

      for (let item in temp) {
        modules.push({title: item, modules: temp[item]});
      }
    }
    this.setState({ modules });
  }

  onChange (event, { newValue, method }) {
    this.setState({
      value: newValue
    });
  }

  onSuggestionsFetchRequested ({ value }) {
    this.setState({
      suggestions: this.getSuggestions(value)
    });
  }

  // Autosuggest will call this function every time you need to clear suggestions.
  onSuggestionsClearRequested () {
    this.setState({
      suggestions: []
    });
  }

  renderSectionTitle (section) {
    return (
      <div>
        <strong>{section.title}</strong>
      </div>
    );
  }

  getSuggestions (value) {
    const escapedValue = this.escapeRegexCharacters(value.trim());

    if (escapedValue === '') {
      return [];
    }

    const regex = new RegExp(escapedValue, 'i');

    return this.state.modules
      .map(section => {
        return {
          title: section.title,
          modules: section.modules.filter(module => regex.test(module.name))
        };
      })
    .filter(section => section.modules.length > 0);
  }

  renderSuggestion (suggestion) {
    return (
      <DraggableModuleSearchResult
        dropModule={this.props.dropModule}
        icon={suggestion.icon}
        name={suggestion.name}
        id={suggestion.id} />
    );
  }

  getSuggestionValue (suggestion) {
    return suggestion.name;
  }

  getSectionSuggestions (section) {
    return section.modules;
  }

  onSuggestionSelected (event, { suggestion }) {
    this.props.addModule(suggestion.id);
    this.setState({value: ''});
  }


  render () {
    const { value, suggestions } = this.state;
    const inputProps = {
      placeholder: 'Search modules',
      value,
      onChange: this.onChange,
      autoFocus: true
    };
    return (
      <Autosuggest
        multiSection={true}
        suggestions={suggestions}
        onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
        onSuggestionsClearRequested={this.onSuggestionsClearRequested}
        getSuggestionValue={this.getSuggestionValue}
        renderSuggestion={this.renderSuggestion}
        renderSectionTitle={this.renderSectionTitle}
        getSectionSuggestions={this.getSectionSuggestions}
        inputProps={inputProps}
        onSuggestionSelected={this.onSuggestionSelected}
      />
    );
  }
}


ModuleSearch.propTypes = {
  addModule:  PropTypes.func.isRequired,
  items:      PropTypes.array.isRequired,
  workflow:   PropTypes.object.isRequired
};
