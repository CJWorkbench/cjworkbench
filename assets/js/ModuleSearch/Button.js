import React from 'react'
import PropTypes from 'prop-types'
import { Manager, Reference } from 'react-popper'
import Popup from './Popup'

export default class Button extends React.PureComponent {
  static propTypes = {
    tabSlug: PropTypes.string.isRequired,
    index: PropTypes.number.isRequired,
    className: PropTypes.string.isRequired,
    isLessonHighlight: PropTypes.bool.isRequired
  }

  state = {
    isOpen: false
  }

  openSearch = () => {
    this.setState({ isOpen: true })
  }

  closeSearch = () => {
    this.setState({ isOpen: false })
  }

  render () {
    const { className, index, tabSlug, isLessonHighlight } = this.props
    const { isOpen } = this.state

    const buttonClassNames = [ 'search' ]
    if (isLessonHighlight) buttonClassNames.push('lesson-highlight')

    return (
      <Manager>
        <Reference>
          {({ ref }) => (
            <div ref={ref} className={className}>
              <button type='button' className={buttonClassNames.join(' ')} onClick={this.openSearch}>
                <i className="icon-addc"></i>{' '}
                <span>ADD STEP</span>
              </button>
            </div>
          )}
        </Reference>
        {isOpen ? (
          <Popup
            index={index}
            tabSlug={tabSlug}
            close={this.closeSearch}
          />
        ) : null}
      </Manager>
    )
  }
}
