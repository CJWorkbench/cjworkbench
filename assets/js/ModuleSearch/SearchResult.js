import React from 'react'
import PropTypes from 'prop-types'
import Popover from 'reactstrap/lib/Popover'
import PopoverHeader from 'reactstrap/lib/PopoverHeader'
import PopoverBody from 'reactstrap/lib/PopoverBody'

export default class SearchResult extends React.PureComponent {
  static propTypes = {
    isActive: PropTypes.bool.isRequired,
    isLessonHighlight: PropTypes.bool.isRequired,
    idName: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    icon: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired, // func(idName) => undefined
    onMouseEnter: PropTypes.func.isRequired, // func(idName) => undefined
  }

  liRef = React.createRef()

  onClick = () => {
    this.props.onClick(this.props.idName)
  }

  onMouseEnter = () => {
    this.props.onMouseEnter(this.props.idName)
  }

  getLiRef = () => {
    // deferred for when Popper asks for it -- not during render, which is too
    // soon.
    return this.liRef.current
  }

  render() {
    const { idName, isActive, isLessonHighlight, isMatch, name, icon, description } = this.props

    const elId = `module-search-result-${idName}`

    const className = [ 'module-search-result' ]
    if (isLessonHighlight) className.push('lesson-highlight')

    return (
      <li ref={this.liRef} className={className.join(' ')} id={elId} data-module-name={name} onClick={this.onClick} onMouseEnter={this.onMouseEnter}>
        <i className={'icon-' + icon}></i>
        <span className='name'>{name}</span>
        {isActive ? (
          <Popover placement='right' isOpen={isActive} target={this.getLiRef} boundariesElement='window' className='module-search-result'>
            <PopoverHeader>{name}</PopoverHeader>
            <PopoverBody>{description}</PopoverBody>
          </Popover>
        ) : null}
      </li>
    )
  }
}
