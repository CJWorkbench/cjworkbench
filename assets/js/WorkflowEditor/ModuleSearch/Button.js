import React from 'react'
import PropTypes from 'prop-types'
import { Manager, Reference } from 'react-popper'
import Popup from './Popup'
import { Trans } from '@lingui/macro'

export default class Button extends React.PureComponent {
  static propTypes = {
    tabSlug: PropTypes.string.isRequired,
    index: PropTypes.number.isRequired,
    className: PropTypes.string.isRequired,
    isLessonHighlight: PropTypes.bool.isRequired,
    isLastAddButton: PropTypes.bool.isRequired
  }

  state = {
    isOpen: false
  }

  handleClick = () => {
    this.setState({ isOpen: true })
  }

  handleClosePopup = () => {
    this.setState({ isOpen: false })
  }

  render () {
    const { className, index, tabSlug, isLessonHighlight, isLastAddButton } = this.props
    const { isOpen } = this.state

    const buttonClassNames = ['search']
    if (isOpen) buttonClassNames.push('active')
    if (isLessonHighlight) buttonClassNames.push('lesson-highlight')

    return (
      <Manager>
        <Reference>
          {({ ref }) => (
            <div ref={ref} className={className}>
              <button type='button' className={buttonClassNames.join(' ')} onClick={this.handleClick}>
                <i className='icon-add' />{' '}
                <span><Trans id='js.WorkflowEditor.ModuleSearch.Button.addStep' description='This should be all-caps for styling reasons'>ADD STEP</Trans></span>
              </button>
            </div>
          )}
        </Reference>
        {isOpen ? (
          <Popup
            isLastAddButton={isLastAddButton}
            index={index}
            tabSlug={tabSlug}
            onClose={this.handleClosePopup}
          />
        ) : null}
      </Manager>
    )
  }
}
