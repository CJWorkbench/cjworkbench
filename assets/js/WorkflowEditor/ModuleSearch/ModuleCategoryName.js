import PropTypes from 'prop-types'
import { Trans } from '@lingui/macro'

export default function ModuleCategoryName (props) {
  const { category } = props
  switch (category) {
    case 'Combine':
      return (
        <Trans id='js.util.ModuleCategoryName.CategoryNames.Combine'>
          Combine
        </Trans>
      )
    case 'Scrape':
      return (
        <Trans id='js.util.ModuleCategoryName.CategoryNames.Scrape'>
          Scrape
        </Trans>
      )
    case 'Clean':
      return (
        <Trans id='js.util.ModuleCategoryName.CategoryNames.Clean'>Clean</Trans>
      )
    case 'Analyze':
      return (
        <Trans id='js.util.ModuleCategoryName.CategoryNames.Analyze'>
          Analyze
        </Trans>
      )
    case 'Visualize':
      return (
        <Trans id='js.util.ModuleCategoryName.CategoryNames.Visualize'>
          Visualize
        </Trans>
      )
    case 'Code':
      return (
        <Trans id='js.util.ModuleCategoryName.CategoryNames.Code'>Code</Trans>
      )
    case 'Add data':
      return (
        <Trans id='js.util.ModuleCategoryName.CategoryNames.AddData'>
          Add data
        </Trans>
      )
    case 'Other':
      return (
        <Trans id='js.util.ModuleCategoryName.CategoryNames.Other'>Other</Trans>
      )
  }
}
ModuleCategoryName.propTypes = {
  category: PropTypes.oneOf([
    'Combine',
    'Scrape',
    'Clean',
    'Analyze',
    'Visualize',
    'Code',
    'Add data',
    'Other'
  ]).isRequired
}
