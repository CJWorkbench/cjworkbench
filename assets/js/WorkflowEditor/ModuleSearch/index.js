import React from 'react';
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { ModulePropType } from './PropTypes'
import Button from './Button'

export default Button
