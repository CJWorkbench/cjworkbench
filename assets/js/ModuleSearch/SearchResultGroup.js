import React from 'react'
import PropTypes from 'prop-types'
import SearchResult from './SearchResult'

export default class ModuleSearchResultGroup extends React.PureComponent {
  static propTypes = {
    name: PropTypes.string.isRequired,
    activeModuleId: PropTypes.number, // null if none active
    modules: PropTypes.arrayOf(PropTypes.shape({
      isLessonHighlight: PropTypes.bool.isRequired,
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      icon: PropTypes.string.isRequired
    })).isRequired,
    onClickModuleId: PropTypes.func.isRequired, // func(moduleId) => undefined
    onMouseEnterModuleId: PropTypes.func.isRequired // func(moduleId) => undefined
  }

  render() {
    const { name, modules, hasMatch, activeModuleId } = this.props

    const children = modules.map(module => (
      <SearchResult
        key={module.name}
        isActive={module.id === activeModuleId}
        {...module}
        onClick={this.props.onClickModuleId}
        onMouseEnter={this.props.onMouseEnterModuleId}
      />
    ))

    return (
      <li className="module-search-result-group" data-name={name}>
        <h4>{name}</h4>
        <ul className="module-search-results">{children}</ul>
      </li>
    )
  }
}
