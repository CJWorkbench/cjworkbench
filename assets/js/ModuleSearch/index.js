/**
* Search field that returns modules matching text input.
*
*/

import React from 'react';
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import SearchResultGroup from './SearchResultGroup'

import lessonSelector from '../lessons/lessonSelector'

const GroupOrder = {
  // dont use 0 -- we use the "||" operator to detect misses
  'Add data': 1,
  'Scrape': 2,
  'Clean': 3,
  'Analyze': 4,
  'Visualize': 5,
  'Code': 6,
}

function compareGroups(a, b) {
  const ai = GroupOrder[a.name] || 99;
  const bi = GroupOrder[b.name] || 99;
  return ai - bi;
}

// Function to sort modules in alphabetical order
function compareModules(a, b) {
  if (a.name.toLowerCase() < b.name.toLowerCase()) return -1
  else if (a.name.toLowerCase() > b.name.toLowerCase()) return 1
  else return 0
}

function groupModules(items) {
  const ret = []
  const temp = {}
  items.sort(compareModules)

  items.forEach(item => {
    if (temp[item.category]) {
      temp[item.category].push(item)
    } else {
      const obj = { name: item.category, modules: [ item ] }
      temp[item.category] = obj.modules
      ret.push(obj)
    }
  })

  ret.sort(compareGroups)

  return ret
}

const escapeRegexCharacters = (str) => {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

export class ModuleSearch extends React.Component {
  static propTypes = {
    modules: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.number.isRequired,
      isLessonHighlight: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      category: PropTypes.string.isRequired,
      icon: PropTypes.string.isRequired,
    })).isRequired,
    index: PropTypes.number.isRequired, // helps mapStateToProps() calculate isLessonHighlight
    isLessonHighlight: PropTypes.bool.isRequired,
    onCancel: PropTypes.func.isRequired, // func() => undefined
    onClickModuleId: PropTypes.func.isRequired, // func(moduleId) => undefined
  }

  state = {
    input: '',
    resultGroups: groupModules(this.props.modules),
    activeModuleId: null
  }

  inputRef = React.createRef()

  findResultGroups(input) {
    const escapedValue = escapeRegexCharacters(input.trim())
    const regex = new RegExp(escapedValue, 'i')
    const foundModules = this.props.modules.filter(m => (regex.test(m.name) | regex.test(m.description)))
    const groups = groupModules(foundModules)
    return groups
  }

  componentDidMount() {
    // auto-focus
    const ref = this.inputRef.current
    if (ref) ref.focus()
  }

  setInput(input) {
    const resultGroups = this.findResultGroups(input)

    this.setState({
      input,
      resultGroups,
    })
  }

  onInputChange = (ev) => {
    const input = ev.target.value
    this.setInput(input)
  }

  onClickModuleId = (moduleId) => {
    this.setInput('')
    this.props.onClickModuleId(moduleId)
  }

  onMouseEnterModuleId = (moduleId) => {
    this.setState({ activeModuleId: moduleId })
  }

  onSubmit = (ev) => {
    ev.preventDefault()
  }

  onReset = () => {
    this.setInput('')
    this.props.onCancel()
  }

  onKeyDown = (ev) => {
    if (ev.keyCode === 27) this.onReset() // Esc => reset
  }

  render () {
    const { input, resultGroups, activeModuleId } = this.state

    const resultGroupComponents = resultGroups.map(rg => (
      <SearchResultGroup
        key={rg.name}
        name={rg.name}
        modules={rg.modules}
        activeModuleId={activeModuleId}
        onClickModuleId={this.onClickModuleId}
        onMouseEnterModuleId={this.onMouseEnterModuleId}
      />
    ))
    const resultGroupsComponent = (
      <ul className="module-search-result-groups">
        {resultGroupComponents}
      </ul>
    )

    const className = [ 'module-search' ]
    if (this.props.isLessonHighlight) className.push('lesson-highlight')

    return (
      <div className={className.join(' ')}>
        <form className="module-search-field" onSubmit={this.onSubmit} onReset={this.onReset}>
          <input
            type='search'
            name='moduleQ'
            placeholder='Searchâ€¦'
            autoComplete='off'
            ref={this.inputRef}
            value={input}
            onChange={this.onInputChange}
            onKeyDown={this.onKeyDown}
            />
          <button type="reset" className="close" title="Close Search"><i className="icon-close"></i></button>
        </form>
        {resultGroupsComponent}
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  const { testHighlight } = lessonSelector(state)
  return {
    isLessonHighlight: testHighlight({ type: 'Module', index: ownProps.index }),
    modules: Object.keys(state.modules).map(moduleId => {
      const module = state.modules[moduleId]
      return {
        ...module,
        isLessonHighlight: testHighlight({ type: 'Module', name: module.name, index: ownProps.index })
      }
    })
  }
}

export default connect(
  mapStateToProps
)(ModuleSearch)
