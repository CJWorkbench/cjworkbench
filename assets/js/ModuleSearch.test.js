/**
 * Testing Stories:
 * -Renders a search bar
 * -Search bar will render suggestions of modules matching input
 * 
 */

import React from 'react'
import ModuleSearch  from './ModuleSearch'
import { mount } from 'enzyme'

describe('ModuleSearch', () => {

  var wrapper;  
  var searchField;
  var addModule = () => {};
  var items = [
    {
      "id":4,
      "name":"Load from Enigma",
      "category":"Add data",
      "description":"Connect a dataset from Enigma's collection via URL.",
      "link":"",
      "author":"Workbench",
      "icon":"url"
    },
    {
      "id":10,
      "name":"Filter by Text",
      "category":"Filter",
      "description":"Filter rows by matching text in specific columns.",
      "link":"",
      "author":"Workbench",
      "icon":"filter"
    }
  ];
  var workflow = {
    "id":15,
    "name":"What a workflow!"
  };

  beforeEach(() => wrapper = mount(
    <ModuleSearch
      addModule={addModule} 
      items={items} 
      workflow={workflow}
    />
  ));
  beforeEach(() => searchField = wrapper.find('.react-autosuggest__input'));    

  it('Renders search bar', () => { 
    expect(wrapper).toMatchSnapshot(); // 1    
  });

  it('Loads modules from props ', (done) => { 
    
    // wait modules to load from props
    setImmediate( () => {
      expect(wrapper).toMatchSnapshot(); 
        
      expect(wrapper.state().modules.length).toBe(2);      

      expect(wrapper.state().modules[0].title).toEqual("Add data");
      expect(wrapper.state().modules[1].title).toEqual("Filter");
      
      done();
    });
  });

  it('Finds a suggestion matching search input', (done) => { 
    // wait modules to load from props
    setImmediate( () => {
      // Search field is focused by default,
      //  enter value to text field
      searchField.simulate('change', {target: {value: 'a'}});
      expect(wrapper).toMatchSnapshot();      
      // check for presence of suggestion
      expect(wrapper.state().suggestions.length).toEqual(1);              
      expect(wrapper.state().suggestions[0].modules[0].name).toEqual("Load from Enigma");      

      done();
    });
  });
    
});

