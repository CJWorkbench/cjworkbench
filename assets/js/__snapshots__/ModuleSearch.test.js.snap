// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleSearch Finds a suggestion matching search input 1`] = `
<ModuleSearch
  addModule={[Function]}
  items={
    Array [
      Object {
        "author": "Workbench",
        "category": "Add data",
        "description": "Connect a dataset from Enigma's collection via URL.",
        "icon": "url",
        "id": 4,
        "link": "",
        "name": "Load from Enigma",
      },
      Object {
        "author": "Workbench",
        "category": "Filter",
        "description": "Filter rows by matching text in specific columns.",
        "icon": "filter",
        "id": 10,
        "link": "",
        "name": "Filter by Text",
      },
    ]
  }
  workflow={
    Object {
      "id": 15,
      "name": "What a workflow!",
    }
  }
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={true}
    getSectionSuggestions={[Function]}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "autoFocus": true,
        "onChange": [Function],
        "placeholder": "Search modules",
        "value": "a",
      }
    }
    multiSection={true}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderSectionTitle={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={
      Array [
        Object {
          "modules": Array [
            Object {
              "author": "Workbench",
              "category": "Add data",
              "description": "Connect a dataset from Enigma's collection via URL.",
              "icon": "url",
              "id": 4,
              "link": "",
              "name": "Load from Enigma",
            },
          ],
          "title": "Add data",
        },
      ]
    }
    theme={
      Object {
        "container": "react-autosuggest__container",
        "containerOpen": "react-autosuggest__container--open",
        "input": "react-autosuggest__input",
        "inputFocused": "react-autosuggest__input--focused",
        "inputOpen": "react-autosuggest__input--open",
        "sectionContainer": "react-autosuggest__section-container",
        "sectionContainerFirst": "react-autosuggest__section-container--first",
        "sectionTitle": "react-autosuggest__section-title",
        "suggestion": "react-autosuggest__suggestion",
        "suggestionFirst": "react-autosuggest__suggestion--first",
        "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
        "suggestionsContainer": "react-autosuggest__suggestions-container",
        "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
        "suggestionsList": "react-autosuggest__suggestions-list",
      }
    }
  >
    <Autowhatever
      getSectionItems={[Function]}
      highlightedItemIndex={null}
      highlightedSectionIndex={null}
      id="1"
      inputProps={
        Object {
          "autoFocus": true,
          "onBlur": [Function],
          "onChange": [Function],
          "onFocus": [Function],
          "onKeyDown": [Function],
          "placeholder": "Search modules",
          "value": "a",
        }
      }
      itemProps={[Function]}
      items={Array []}
      multiSection={true}
      renderInputComponent={[Function]}
      renderItem={[Function]}
      renderItemData={
        Object {
          "query": "a",
        }
      }
      renderItemsContainer={[Function]}
      renderSectionTitle={[Function]}
      theme={
        Object {
          "container": "react-autosuggest__container",
          "containerOpen": "react-autosuggest__container--open",
          "input": "react-autosuggest__input",
          "inputFocused": "react-autosuggest__input--focused",
          "inputOpen": "react-autosuggest__input--open",
          "item": "react-autosuggest__suggestion",
          "itemFirst": "react-autosuggest__suggestion--first",
          "itemHighlighted": "react-autosuggest__suggestion--highlighted",
          "itemsContainer": "react-autosuggest__suggestions-container",
          "itemsContainerOpen": "react-autosuggest__suggestions-container--open",
          "itemsList": "react-autosuggest__suggestions-list",
          "sectionContainer": "react-autosuggest__section-container",
          "sectionContainerFirst": "react-autosuggest__section-container--first",
          "sectionTitle": "react-autosuggest__section-title",
        }
      }
    >
      <div
        className="react-autosuggest__container"
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-expanded={false}
          aria-haspopup={false}
          aria-owns="react-autowhatever-1"
          autoComplete="off"
          autoFocus={true}
          className="react-autosuggest__input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Search modules"
          role="combobox"
          type="text"
          value="a"
        />
        <div
          className="react-autosuggest__suggestions-container"
          id="react-autowhatever-1"
        />
      </div>
    </Autowhatever>
  </Autosuggest>
</ModuleSearch>
`;

exports[`ModuleSearch Loads modules from props  1`] = `
<ModuleSearch
  addModule={[Function]}
  items={
    Array [
      Object {
        "author": "Workbench",
        "category": "Add data",
        "description": "Connect a dataset from Enigma's collection via URL.",
        "icon": "url",
        "id": 4,
        "link": "",
        "name": "Load from Enigma",
      },
      Object {
        "author": "Workbench",
        "category": "Filter",
        "description": "Filter rows by matching text in specific columns.",
        "icon": "filter",
        "id": 10,
        "link": "",
        "name": "Filter by Text",
      },
    ]
  }
  workflow={
    Object {
      "id": 15,
      "name": "What a workflow!",
    }
  }
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={true}
    getSectionSuggestions={[Function]}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "autoFocus": true,
        "onChange": [Function],
        "placeholder": "Search modules",
        "value": "",
      }
    }
    multiSection={true}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderSectionTitle={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
    theme={
      Object {
        "container": "react-autosuggest__container",
        "containerOpen": "react-autosuggest__container--open",
        "input": "react-autosuggest__input",
        "inputFocused": "react-autosuggest__input--focused",
        "inputOpen": "react-autosuggest__input--open",
        "sectionContainer": "react-autosuggest__section-container",
        "sectionContainerFirst": "react-autosuggest__section-container--first",
        "sectionTitle": "react-autosuggest__section-title",
        "suggestion": "react-autosuggest__suggestion",
        "suggestionFirst": "react-autosuggest__suggestion--first",
        "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
        "suggestionsContainer": "react-autosuggest__suggestions-container",
        "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
        "suggestionsList": "react-autosuggest__suggestions-list",
      }
    }
  >
    <Autowhatever
      getSectionItems={[Function]}
      highlightedItemIndex={null}
      highlightedSectionIndex={null}
      id="1"
      inputProps={
        Object {
          "autoFocus": true,
          "onBlur": [Function],
          "onChange": [Function],
          "onFocus": [Function],
          "onKeyDown": [Function],
          "placeholder": "Search modules",
          "value": "",
        }
      }
      itemProps={[Function]}
      items={Array []}
      multiSection={true}
      renderInputComponent={[Function]}
      renderItem={[Function]}
      renderItemData={
        Object {
          "query": "",
        }
      }
      renderItemsContainer={[Function]}
      renderSectionTitle={[Function]}
      theme={
        Object {
          "container": "react-autosuggest__container",
          "containerOpen": "react-autosuggest__container--open",
          "input": "react-autosuggest__input",
          "inputFocused": "react-autosuggest__input--focused",
          "inputOpen": "react-autosuggest__input--open",
          "item": "react-autosuggest__suggestion",
          "itemFirst": "react-autosuggest__suggestion--first",
          "itemHighlighted": "react-autosuggest__suggestion--highlighted",
          "itemsContainer": "react-autosuggest__suggestions-container",
          "itemsContainerOpen": "react-autosuggest__suggestions-container--open",
          "itemsList": "react-autosuggest__suggestions-list",
          "sectionContainer": "react-autosuggest__section-container",
          "sectionContainerFirst": "react-autosuggest__section-container--first",
          "sectionTitle": "react-autosuggest__section-title",
        }
      }
    >
      <div
        className="react-autosuggest__container"
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-expanded={false}
          aria-haspopup={false}
          aria-owns="react-autowhatever-1"
          autoComplete="off"
          autoFocus={true}
          className="react-autosuggest__input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Search modules"
          role="combobox"
          type="text"
          value=""
        />
        <div
          className="react-autosuggest__suggestions-container"
          id="react-autowhatever-1"
        />
      </div>
    </Autowhatever>
  </Autosuggest>
</ModuleSearch>
`;

exports[`ModuleSearch Renders search bar 1`] = `
<ModuleSearch
  addModule={[Function]}
  items={
    Array [
      Object {
        "author": "Workbench",
        "category": "Add data",
        "description": "Connect a dataset from Enigma's collection via URL.",
        "icon": "url",
        "id": 4,
        "link": "",
        "name": "Load from Enigma",
      },
      Object {
        "author": "Workbench",
        "category": "Filter",
        "description": "Filter rows by matching text in specific columns.",
        "icon": "filter",
        "id": 10,
        "link": "",
        "name": "Filter by Text",
      },
    ]
  }
  workflow={
    Object {
      "id": 15,
      "name": "What a workflow!",
    }
  }
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={true}
    getSectionSuggestions={[Function]}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "autoFocus": true,
        "onChange": [Function],
        "placeholder": "Search modules",
        "value": "",
      }
    }
    multiSection={true}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderSectionTitle={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
    theme={
      Object {
        "container": "react-autosuggest__container",
        "containerOpen": "react-autosuggest__container--open",
        "input": "react-autosuggest__input",
        "inputFocused": "react-autosuggest__input--focused",
        "inputOpen": "react-autosuggest__input--open",
        "sectionContainer": "react-autosuggest__section-container",
        "sectionContainerFirst": "react-autosuggest__section-container--first",
        "sectionTitle": "react-autosuggest__section-title",
        "suggestion": "react-autosuggest__suggestion",
        "suggestionFirst": "react-autosuggest__suggestion--first",
        "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
        "suggestionsContainer": "react-autosuggest__suggestions-container",
        "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
        "suggestionsList": "react-autosuggest__suggestions-list",
      }
    }
  >
    <Autowhatever
      getSectionItems={[Function]}
      highlightedItemIndex={null}
      highlightedSectionIndex={null}
      id="1"
      inputProps={
        Object {
          "autoFocus": true,
          "onBlur": [Function],
          "onChange": [Function],
          "onFocus": [Function],
          "onKeyDown": [Function],
          "placeholder": "Search modules",
          "value": "",
        }
      }
      itemProps={[Function]}
      items={Array []}
      multiSection={true}
      renderInputComponent={[Function]}
      renderItem={[Function]}
      renderItemData={
        Object {
          "query": "",
        }
      }
      renderItemsContainer={[Function]}
      renderSectionTitle={[Function]}
      theme={
        Object {
          "container": "react-autosuggest__container",
          "containerOpen": "react-autosuggest__container--open",
          "input": "react-autosuggest__input",
          "inputFocused": "react-autosuggest__input--focused",
          "inputOpen": "react-autosuggest__input--open",
          "item": "react-autosuggest__suggestion",
          "itemFirst": "react-autosuggest__suggestion--first",
          "itemHighlighted": "react-autosuggest__suggestion--highlighted",
          "itemsContainer": "react-autosuggest__suggestions-container",
          "itemsContainerOpen": "react-autosuggest__suggestions-container--open",
          "itemsList": "react-autosuggest__suggestions-list",
          "sectionContainer": "react-autosuggest__section-container",
          "sectionContainerFirst": "react-autosuggest__section-container--first",
          "sectionTitle": "react-autosuggest__section-title",
        }
      }
    >
      <div
        className="react-autosuggest__container"
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-expanded={false}
          aria-haspopup={false}
          aria-owns="react-autowhatever-1"
          autoComplete="off"
          autoFocus={true}
          className="react-autosuggest__input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Search modules"
          role="combobox"
          type="text"
          value=""
        />
        <div
          className="react-autosuggest__suggestions-container"
          id="react-autowhatever-1"
        />
      </div>
    </Autowhatever>
  </Autosuggest>
</ModuleSearch>
`;
